peps/pep-0756.rst

```diff
@@ -0,0 +1,359 @@
+An exported string does not end with a trailing NUL character: the
+``PyUnicode_Export()`` caller must use ``Py_buffer.len`` to get the
+string length.
+
+
+Implementation
+==============
+
+https://github.com/python/cpython/pull/123738
+
+
+Backwards Compatibility
+=======================
+
+There is no impact on the backward compatibility, only new C API
+functions are added.
+
+
+Usage of PEP 393 C APIs
+=======================
+
+A code search on PyPI top 7,500 projects (on March 2024) shows that
+there are many projects importing and exporting UCS formats with the
+regular C API.
+
+PyUnicode_FromKindAndData()
+---------------------------
+
+25 projects call ``PyUnicode_FromKindAndData()``:
+
+* **Cython** (3.0.9)
+* Levenshtein (0.25.0)
+* PyICU (2.12)
+* PyICU-binary (2.7.4)
+* PyQt5 (5.15.10)
+* PyQt6 (6.6.1)
+* aiocsv (1.3.1)
+* asyncpg (0.29.0)
+* biopython (1.83)
+* catboost (1.2.3)
+* cffi (1.16.0)
+* mojimoji (0.0.13)
+* mwparserfromhell (0.6.6)
+* numba (0.59.0)
+* **numpy** (1.26.4)
+* orjson (3.9.15)
+* pemja (0.4.1)
+* pyahocorasick (2.0.0)
+* pyjson5 (1.6.6)
+* rapidfuzz (3.6.2)
+* regex (2023.12.25)
+* srsly (2.4.8)
+* tokenizers (0.15.2)
+* ujson (5.9.0)
+* unicodedata2 (15.1.0)
+
+
+PyUnicode_4BYTE_DATA()
+----------------------
+
+21 projects call ``PyUnicode_2BYTE_DATA()`` and/or
+``PyUnicode_4BYTE_DATA()``:
+
+* **Cython** (3.0.9)
+* **MarkupSafe** (2.1.5)
+* Nuitka (2.1.2)
+* PyICU (2.12)
+* PyICU-binary (2.7.4)
+* PyQt5_sip (12.13.0)
+* PyQt6_sip (13.6.0)
+* biopython (1.83)
+* catboost (1.2.3)
+* cement (3.0.10)
+* cffi (1.16.0)
+* duckdb (0.10.0)
+* **mypy** (1.9.0)
+* **numpy** (1.26.4)
+* orjson (3.9.15)
+* pemja (0.4.1)
+* pyahocorasick (2.0.0)
+* pyjson5 (1.6.6)
+* pyobjc-core (10.2)
+* sip (6.8.3)
+* wxPython (4.2.1)
+
+
+Rejected Ideas
+==============
+
+Reject embedded NUL character and trailing NUL character
+--------------------------------------------------------
+
+In C, it's convenient to have a trailing NUL character. For example,
+the ``for (; *str != 0; str++)`` loop can be used to iterate on
+characters and ``strlen()`` can be used to get a string length.
+
+The problem is that a Python ``str`` object can embed NUL characters.
+Example: ``"ab\0c"``. If a string contains an embedded NUL character,
+code relying on the NUL character to find the string end truncate the
+string. It can lead to bugs, or even security vulnerabilities.
+
+Rejecting embedded NUL characters require to scan the string which has
+a O(n) complexity.
+
+Reject surrogate characters
+---------------------------
+
+Surrogate characters are characters in the Unicode range [U+D800;
+U+DFFF].  They are disallowed by UTF codecs such as UTF-8. A Python
+``str`` object can contain arbitrary lone surrogate characters. Example:
+``"\uDC80"``.
+
+Rejecting surogate characters prevents exporting a string which contain
+such character. It can be surprising and annoying since the
+``PyUnicode_Export()`` caller doesn't control the string content.
```

Comment:
```suggestion
Rejecting surrogate characters prevents exporting a string which contains
such a character. It can be surprising and annoying since the
``PyUnicode_Export()`` caller doesn't control the string contents.
```